bundles/org.smarthomej.binding.viessmann/src/main/java/org/smarthomej/binding/viessmann/internal/api/ViessmannAuth.java

```diff
@@ -0,0 +1,326 @@
+            String user, String password) {
+        this.api = api;
+        this.apiKey = apiKey;
+        this.httpClient = httpClient;
+        this.bridgeHandler = bridgeHandler;
+        this.user = user;
+        this.password = password;
+        state = ViessmannAuthState.NEED_AUTH;
+        authResponse = null;
+    }
+
+    public void setState(ViessmannAuthState newState) {
+        if (newState != state) {
+            logger.debug("ViessmannAuth: Change state from {} to {}", state, newState);
+            state = newState;
+        }
+    }
+
+    public void setRefreshToken(String newRefreshToken) {
+        if (!newRefreshToken.equals(refreshToken)) {
+            logger.debug("ViessmannAuth: Change refreshToken from {} to {}", refreshToken, newRefreshToken);
+            refreshToken = newRefreshToken;
+        }
+    }
+
+    public boolean isComplete() {
+        return state == ViessmannAuthState.COMPLETE;
+    }
+
+    public ViessmannAuthState doAuthorization() throws ViessmannAuthException {
+        switch (state) {
+            case NEED_AUTH:
+                authorize();
+                if (state == ViessmannAuthState.NEED_TOKEN) {
+                    getTokens();
+                }
+                break;
+            case NEED_LOGIN:
+                break;
+            case NEED_TOKEN:
+                getTokens();
+                break;
+            case NEED_REFRESH_TOKEN:
+                getRefreshTokens();
+                break;
+            case COMPLETE:
+                bridgeHandler.updateBridgeStatus(ThingStatus.ONLINE);
+                break;
+        }
+        return state;
+    }
+
+    public @Nullable String getAccessToken() throws ViessmannAuthException {
+        return this.accessToken;
+    }
+
+    public void setAccessToken(String newAccessToken) {
+        this.accessToken = newAccessToken;
+    }
+
+    /**
+     * Call the Viessmann authorize endpoint to get the authorization code.
+     */
+    private void authorize() throws ViessmannAuthException {
+        logger.debug("ViessmannAuth: State is {}: Executing step: 'authorize'", state);
+        StringBuilder url = new StringBuilder(VIESSMANN_AUTHORIZE_URL);
+        url.append("?response_type=code");
+        url.append("&client_id=").append(apiKey);
+        url.append("&code_challenge=2e21faa1-db2c-4d0b-a10f-575fd372bc8c-575fd372bc8c");
+        url.append("&redirect_uri=http://localhost:8080/viessmann/authcode/");
+        url.append("&scope=").append(VIESSMANN_SCOPE);
+        logger.trace("ViessmannAuth: Getting authorize URL={}", url);
+        String response = executeUrlAuthorize("GET", url.toString());
+        logger.trace("ViessmannAuth: Auth response: {}", response);
+        if (response != null) {
+            if (response.indexOf("<!DOCTYPE html>") >= 0) {
+                logger.error("ViessmannAuth: Login failed. Please check user and passowrd.");
```

Comment:
```suggestion
                logger.warn("ViessmannAuth: Login failed. Please check user and password.");
```

Never log at `ERROR` level, except the failure is so bad that the stability of the whole system is in danger. Also below.